### PR TITLE
Add lowering of All and Any intrinsics

### PR DESCRIPTION
**Pull request by mleair**

### CODE REVIEW

**Comment by mleair** [comment]
Please note: I was only able to test the case where all/any returns a scalar (i.e., when the mask argument is rank 1 or there is no dim argument). The case where all/any returns an array, I get the following TO DO message out of bbc:

ConvertExpr.cpp:2248: not yet implemented non-elemental intrinsic ref

I believe this falls under the area of array assignment which is a work in progress.

File: flang/include/flang/Lower/ReductionRuntime.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#define FORTRAN_LOWER_REDUCTIONRUNTIME_H
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+
+namespace fir {
+class ExtendedValue;
+}
+
+namespace Fortran {
+namespace lower {
```
Comment (schweitzpgi) [inline flang/include/flang/Lower/ReductionRuntime.h:19]: You can combine these as `namespace Fortran::lower {`

File: flang/lib/Lower/ReductionRuntime.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+using namespace Fortran::runtime;
+
+/// Generate calls to reduction intrinsics such as All and Any.
+/// These are the descriptor based implementations that take two 
+/// arguments (mask, dim).
+template <typename FN>
+static void
+genRed2Args(FN func, Fortran::lower::FirOpBuilder &builder,
+            mlir::Location loc, mlir::Value resultBox,
+            mlir::Value maskBox, mlir::Value dim);
```
Comment (schweitzpgi) [inline flang/lib/Lower/ReductionRuntime.cpp:28]: It doesn't look like you need to forward declare this template but can just move the definition to here.

File: flang/test/Lower/intrinsic-procedures.f90 [hunk 3]
```diff
@@ -70,6 +82,18 @@ subroutine anint_test(a, b)
...
+! ANY
+! CHECK-LABEL: any_test
+! CHECK-SAME: %[[arg0:.*]]: !fir.box<!fir.array<?x!fir.logical<4>>>) -> !fir.logical<4>
+logical function any_test(mask)
+  logical :: mask(:)
+! CHECK: %[[c1:.*]] = constant 1 : index
+! CHECK: %[[a1:.*]] = fir.convert %[[arg0]] : (!fir.box<!fir.array<?x!fir.logical<4>>>) -> !fir.box<none>
+! CHECK: %[[a2:.*]] = fir.convert %[[c1]] : (index) -> i32
+  any_test = any(mask)
+! CHECK:  %[[a3:.*]] = fir.call @_FortranAAny(%[[a1]], %{{.*}}, %[[a2]]) : (!fir.box<none>, !fir.ref<i8>, i32, i32) -> i1
```
Comment (schweitzpgi) [inline flang/test/Lower/intrinsic-procedures.f90:94]: There are 4 arguments, but the regular expression appears to only be looking for 3. (Two get matched with the single `%{{.*}},` in the middle.) That's a bit confusing at first glance.

File: flang/test/Lower/intrinsic-procedures.f90 [hunk 4]
```diff
@@ -44,6 +44,18 @@ subroutine aint_test(a, b)
...
+! ALL
+! CHECK-LABEL: all_test
+! CHECK-SAME: %[[arg0:.*]]: !fir.box<!fir.array<?x!fir.logical<4>>>) -> !fir.logical<4>
+logical function all_test(mask)
+  logical :: mask(:)
+! CHECK: %[[c1:.*]] = constant 1 : index
+! CHECK: %[[a1:.*]] = fir.convert %[[arg0]] : (!fir.box<!fir.array<?x!fir.logical<4>>>) -> !fir.box<none>
+! CHECK: %[[a2:.*]] = fir.convert %[[c1]] : (index) -> i32
+  all_test = all(mask)
+! CHECK:  %[[a3:.*]] = fir.call @_FortranAAll(%[[a1]], %{{.*}}, %[[a2]]) : (!fir.box<none>, !fir.ref<i8>, i32, i32) -> i1
```
Comment (schweitzpgi) [inline flang/test/Lower/intrinsic-procedures.f90:56]: nit: two for one match. See below.

File: flang/lib/Lower/ReductionRuntime.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+/// Generate calls to reduction intrinsics such as All and Any.
+/// These are the descriptor based implementations that take two 
+/// arguments (mask, dim).
+template <typename FN>
+static void
+genRed2Args(FN func, Fortran::lower::FirOpBuilder &builder,
+            mlir::Location loc, mlir::Value resultBox,
+            mlir::Value maskBox, mlir::Value dim) {
+
+
```
Comment (schweitzpgi) [inline flang/lib/Lower/ReductionRuntime.cpp:67]: nit: why 2 blanks lines? why any blank lines?

File: flang/test/Lower/intrinsic-procedures.f90 (hunk 3 above)
Comment (mleair) [reply flang/test/Lower/intrinsic-procedures.f90:94]: The runtime routine's arguments are:

const Descriptor &x, const char *source, int line, int dim

The two arguments that are being "wild carded" are the source info arguments that we usually "wild card" from what I've seen in other tests. The first argument is the mask argument and the last argument is the optional dim argument.

Comment (mleair) [reply flang/test/Lower/intrinsic-procedures.f90:94]: Let me know if there's a better way to write this.

File: flang/lib/Lower/ReductionRuntime.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                                 mlir::Location loc,
+                                 mlir::Value resultBox, mlir::Value maskBox,
+                                 mlir::Value dim) {
+  auto allFunc = Fortran::lower::getRuntimeFunc<mkRTKey(AllDim)>(loc, builder);
+  genRed2Args(allFunc, builder, loc, resultBox, maskBox, dim);
+}
+
+
+/// Generate call to any runtime routine.
+/// This calls the descriptor based runtime call implementation of the scan
```
Comment (jeanPerier) [inline flang/lib/Lower/ReductionRuntime.cpp:59]: Nit: `the scan` -> `any`

File: flang/lib/Lower/ReductionRuntime.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    builder.createConvert(loc, fTy.getInput(2), dim),
+    builder.createConvert(loc, fTy.getInput(3), sourceFile),
+    builder.createConvert(loc, fTy.getInput(4), sourceLine) 
+  };
+   
+  builder.create<fir::CallOp>(loc, func, args);
+
+}
+/// Generate call to all runtime routine.
+/// This calls the descriptor based runtime call implementation of the scan
```
Comment (jeanPerier) [inline flang/lib/Lower/ReductionRuntime.cpp:46]: Nit: `the scan` -> `all`

File: flang/include/flang/Lower/ReductionRuntime.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace fir {
+class ExtendedValue;
+}
+
+namespace Fortran::lower {
+class FirOpBuilder;
+
+/// Generate call to all runtime routine.
+/// This calls the descriptor based runtime call implementation of the scan
```
Comment (jeanPerier) [inline flang/include/flang/Lower/ReductionRuntime.h:22]: Nit: `the scan` -> all

File: flang/include/flang/Lower/ReductionRuntime.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+/// intrinsics.
+void genAllDescriptor(Fortran::lower::FirOpBuilder &builder, 
+                         mlir::Location loc,
+                         mlir::Value resultBox, mlir::Value maskBox,
+                         mlir::Value dim); 
+
+/// Generate call to any runtime routine.
+/// This calls the descriptor based runtime call implementation of the scan
```
Comment (jeanPerier) [inline flang/include/flang/Lower/ReductionRuntime.h:30]: Nit: `the scan` -> any

File: flang/test/Lower/intrinsic-procedures.f90 (hunk 3 above)
Comment (schweitzpgi) [reply flang/test/Lower/intrinsic-procedures.f90:94]: I was thinking something like `%{{.*}}, %{{.*}},` would make it clearer that you expect 2 distinct arguments.

**Comment by mleair** [comment]
PR #733 superscedes this.

**Pull request closed without merging**